src/type_seed.jl

```diff
@@ -0,0 +1,76 @@
+"""
+    type_seed(x)
+
+Computes a value to use as a seed for computing the hash value of a type.
+"""
+function type_seed end
+
+function type_seed(m::Module, h::UInt)
+    if m === Main || m === Base || m === Core
+        return h
+    else
+        mp = parentmodule(m)
+        if mp === m
+            return h
+        else
+            return hash(nameof(m), type_seed(mp, h))
+        end
+    end
+end
+
+function type_seed(t::UnionAll, h::UInt)
+    h = hash(h, 0xbd5f3e4941dba79d)
+    h = type_seed(t.var, h)
+    h = type_seed(t.body, h)
+    h
+end
+
+function type_seed(x::Union, h::UInt)
+    h = hash(h, 0xa16a31201c4852c2)
+    for t in Base.uniontypes(x)
+        h = type_seed(t, h)
+    end
+    return h
+end
+
+function type_seed(t::TypeVar, h::UInt)
```

Comment:
This works because types in Julia cannot have recursive bounds:

```julia
julia> abstract type R{T <: R{T}}; end
ERROR: UndefVarError: R not defined
Stacktrace:
 [1] top-level scope
   @ REPL[55]:1

julia> abstract type R{T}; end

julia> struct S{T <: R{T}}; end
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] top-level scope
   @ REPL[57]:1
```